src/bvals/cc/bvals_cc_in_one.cpp

```diff
@@ -0,0 +1,553 @@
+    if (nb.ni.fi1 == 1)
+      si += pmb->block_size.nx1 / 2;
+    else
+      ei -= pmb->block_size.nx1 / 2;
+  } else if (nb.ni.ox1 > 0) {
+    si = cellbounds.ie(interior) + 1;
+    ei = cellbounds.ie(interior) + NGHOST;
+  } else {
+    si = cellbounds.is(interior) - NGHOST;
+    ei = cellbounds.is(interior) - 1;
+  }
+
+  if (nb.ni.ox2 == 0) {
+    sj = cellbounds.js(interior);
+    ej = cellbounds.je(interior);
+    if (pmb->block_size.nx2 > 1) {
+      if (nb.ni.ox1 != 0) {
+        if (nb.ni.fi1 == 1)
+          sj += pmb->block_size.nx2 / 2;
+        else
+          ej -= pmb->block_size.nx2 / 2;
+      } else {
+        if (nb.ni.fi2 == 1)
+          sj += pmb->block_size.nx2 / 2;
+        else
+          ej -= pmb->block_size.nx2 / 2;
+      }
+    }
+  } else if (nb.ni.ox2 > 0) {
+    sj = cellbounds.je(interior) + 1;
+    ej = cellbounds.je(interior) + NGHOST;
+  } else {
+    sj = cellbounds.js(interior) - NGHOST;
+    ej = cellbounds.js(interior) - 1;
+  }
+
+  if (nb.ni.ox3 == 0) {
+    sk = cellbounds.ks(interior);
+    ek = cellbounds.ke(interior);
+    if (pmb->block_size.nx3 > 1) {
+      if (nb.ni.ox1 != 0 && nb.ni.ox2 != 0) {
+        if (nb.ni.fi1 == 1)
+          sk += pmb->block_size.nx3 / 2;
+        else
+          ek -= pmb->block_size.nx3 / 2;
+      } else {
+        if (nb.ni.fi2 == 1)
+          sk += pmb->block_size.nx3 / 2;
+        else
+          ek -= pmb->block_size.nx3 / 2;
+      }
+    }
+  } else if (nb.ni.ox3 > 0) {
+    sk = cellbounds.ke(interior) + 1;
+    ek = cellbounds.ke(interior) + NGHOST;
+  } else {
+    sk = cellbounds.ks(interior) - NGHOST;
+    ek = cellbounds.ks(interior) - 1;
+  }
+}
+
+//----------------------------------------------------------------------------------------
+//! \fn void cell_centered_bvars::CalcIndicesLoadSame(int ox, int &s, int &e,
+//                                                    const IndexRange &bounds)
+//  \brief Calculate indices for LoadBoundary routines for buffers on the same level
+//         and to coarser.
+
+void CalcIndicesLoadSame(int ox, int &s, int &e, const IndexRange &bounds) {
+  if (ox == 0) {
+    s = bounds.s;
+    e = bounds.e;
+  } else if (ox > 0) {
+    s = bounds.e - NGHOST + 1;
+    e = bounds.e;
+  } else {
+    s = bounds.s;
+    e = bounds.s + NGHOST - 1;
+  }
+}
+
+//----------------------------------------------------------------------------------------
+//! \fn void cell_centered_bvars::CalcIndicesLoadToFiner(int &si, int &ei, int &sj,
+//                                                       int &ej, int &sk, int &ek,
+//                                                       const NeighborBlock &nb,
+//                                                       MeshBlock *pmb)
+//  \brief Calculate indices for LoadBoundary routines for buffers to finer levels
+
+void CalcIndicesLoadToFiner(int &si, int &ei, int &sj, int &ej, int &sk, int &ek,
+                            const NeighborBlock &nb, MeshBlock *pmb) {
```

Comment:
These are pretty much always the same fore each combination of coarse/fine/relative orientation. Maybe we should pre-compute them once and cache them.

Reply:
True. I think that's a nice to have for a future PR. Given that the computed indices are already cached and that this operation in generally is not that expensive, I don't think it's a performance critical part right now.

Reply:
Sure, true enough. Precomputing them once might be less bug prone than checking for allocation, however.

Reply:
There a subtle difference here.
The "checking for allocation" has to be done independent of  the caching here, because after each refinement, derefinement, or loadbalancing step the cached `boundary_info` is not valid any more[*] whereas these indices are fixed for a given combination neighbor configuration.
For example, right face has a coarse neighbor and there are 2 ghost cells would always yield identical indices for this "configuration", but whether the right face has a coarse neighbor may change between steps.

[*] Actually, this may be where the current caching fails. Let me double check if the buffers are only invalidated for load balancing or also for refinement/derefinement (without load balancing).
In the latter case, we'd need to clear the cache and then this should work.

Reply:
> [*] Actually, this may be where the current caching fails. Let me double check if the buffers are only invalidated for load balancing or also for refinement/derefinement (without load balancing).
> In the latter case, we'd need to clear the cache and then this should work.

As far as I can tell that's not it. Clearing the caches is always called.

Reply:
> For example, right face has a coarse neighbor and there are 2 ghost cells would always yield identical indices for this "configuration", but whether the right face has a coarse neighbor may change between steps.

Ah, I see. Yes, this is true.